### PR TITLE
SITL: Add Rover vehicle for AirSim

### PR DESCRIPTION
**Pull request by rajat2004**
[AirSim](https://github.com/microsoft/AirSim) has a Car vehicle and this PR adds support for using the same for Rover SITL
Small video- https://drive.google.com/open?id=1MgqXYT7ouC9LMz35czZfZ8l2pyW5bXRF

AirSim branch - https://github.com/rajat2004/AirSim/tree/pr-ardurover2
(Will be opening PR soon)

Major blockage (Not occuring now, but will be good to test on different systems) - Mavproxy disconnects after 10-15 seconds, my workaround was to use `--no-mavproxy` option in `sim_vehicle.py` and then start Mavproxy from a different terminal, it shows Link Down in the beginning but connects after 1-2 secs and then works properly

Note that Car vehicle in AirSim uses wall-clock and doesn't have a fixed timestep, so timing will be a problem

### CODE REVIEW

**Comment by rajat2004** [comment]
A log file - https://drive.google.com/open?id=1P_u7l_5v53gD_-v0G143i7_XgLQo7BlX
Wireshark capture - https://drive.google.com/open?id=1VeCUAK-Ks7DwK3ITRqlYSNyTQonGwxcq

The messages are still coming, but the deltat is vey high

File: libraries/SITL/SIM_AirSim.cpp [hunk 1]
```diff
@@ -281,7 +308,7 @@ void AirSim::recv_fdm()
         rcin[i] = state.rc.rc_channels.data[i];
     }
 
-#if 0
+// #if 0
```
Comment (rajat2004) [inline libraries/SITL/SIM_AirSim.cpp]: Commented out for testing, will be removed when ready to merge

**Comment by khancyr** [comment]
Check vibration and gyro. On gazebo, it get the same issue on start when doing gyro calibration get long because of ground collision model

**Comment by rajat2004** [comment]
@khancyr Thanks for the tip!
There was a problem with the sensor data being inconsistent due to a small offset from AirSim, but fixed that

Right now the Gyro calibration is disabled by setting `INS_GYR_CAL` to 0, since it was stuck on the calibration earlier
Will check the sensor data for vibration and gyro

**Comment by rajat2004** [comment]
Didn't get much time for checking the sensor data, instead cleaned up the AirSim code and opened an intermediate PR for the restructuring of Airsim's Car code, since that could take a longer time to get merged
https://github.com/microsoft/AirSim/pull/2172

**Comment by rajat2004** [comment]
@khancyr I had tried various things over the week but nothing worked. The calibration was also disabled using `INS_GYR_CAL` and also tried other Crash checks params and few others, I'll put some more details in some time

Any idea about where else it could be going wrong? One suspicious point is that it works if I start mavproxy separately

**Comment by rajat2004** [comment]
@khancyr I found this issue created by you- https://github.com/ArduPilot/ardupilot/issues/5615
Could you give some more info on how you fixed that?

**Comment by khancyr** [comment]
hello,
I was wrong, my problem was different at it appears at gyro calibration, your append after. Attach GDB to SITL to see where it crash

**Comment by rajat2004** [comment]
@khancyr Did that, no crash!

**Comment by rajat2004** [comment]
Been quite some time, sorry for that!
The first AirSim PR needed changes after rebase and review was done very recently, so hopefully, that goes in soon.

I still haven't been able to figure out what's causing the slowdown on Ardupilot side to send the GCS messages, innovations go off after console shows Link Down, not good in Log Analysis but am trying to figure out as much as possible

2 logs, logging enabled when disarmed - [EKF Type 2](https://drive.google.com/open?id=15XCKyXNWEe7yH0kYOKB-uutooqj5ibgJ), [EKF Type 10](https://drive.google.com/open?id=1TR-GVNqZYAj9q9dYOMSI2JDcnJkKi4UG)

Any help would be great

**Comment by rajat2004** [comment]
Hi,
So I'm putting the instructions on how to use the ArduRover support with AirSim here, will get back to this as soon as possible.

All this has been tested on Ubuntu only till now.

First, for the AirSim branch, use this one- https://github.com/rajat2004/AirSim/tree/pr-ardurover2
I'll try to update it to match with the latest AirSim master and some other changes but will take a few days.

Next, the `settings.json` for ArduRover:
```
{
  "SettingsVersion": 1.2,
  "LocalHostIp": "127.0.0.1",
  "SimMode": "Car",
  "OriginGeopoint": {
    "Latitude": -35.363261,
    "Longitude": 149.165230,
    "Altitude": 583
  },
  "Vehicles": {
    "Rover": {
      "VehicleType": "ArduRover",
      "UseSerial": false,
      "DefaultVehicleState": "Disarmed",
      "UdpIp": "127.0.0.1",
      "UdpPort": 9003,
      "SitlPort": 9002,
      "AutoCreate": true,
      "Sensors": {
        "Imu": {
          "SensorType": 2,
          "Enabled": true
        },
        "Gps": {
          "SensorType": 3,
          "Enabled": true
        }
      }
    }
  }
}
```

The SITL commands:
```
sim_vehicle.py -f airsim-rover -w --add-param-file=../libraries/SITL/examples/Airsim/rover.parm --no-mavproxy
```
To be ran from the `APMRover2` directory, if running from the `ardupilot` root directory, add `-v APMRover2` and change the location of the param file to `libraries/SITL/examples/Airsim/rover.parm`

Now, normally we use the `--console --map` to bring the MAVProxy modules, but MAVProxy disconnects after some 10-15 seconds, therefore the workaround which works for me is to first add the `--no-mavproxy` option which is there in the command above.

After running `sim_vehicle.py`, start MAVProxy (Directly taken from the output show by `sim_vehicle.py` when started normally) :
```
mavproxy.py --master tcp:127.0.0.1:5760 --sitl 127.0.0.1:5501 --out 127.0.0.1:14550 --out 127.0.0.1:14551 --console --map
```

The console still shows `Link Down` but it goes away after a few seconds and everything works as expected after that

**Comment by peterbarker** [comment]
@Hwurzburg for visibility; we may need to swipe some of this into the Wiki, depending on when this is merged...

**Comment by rajat2004** [comment]
@peterbarker Thanks for this, my main intention of the above post was to add the instructions so that others can try it, someone recently asked me for them. Another was to have a record since I'll my reinstalling my system soon and will need them eventually.

I'll create a PR for the Wiki once this get's merged.

@peterbarker @khancyr @tridge @rmackay9 
I still haven't been able to figure out what's causing the MAVProxy disconnection problem, any pointers would be great to have

**Comment by rajat2004** [comment]
Things are working on my system right now, without the MAVProxy workaround
Tuning is needed to make it's performance better
Will post Logs soon

**Comment by rajat2004** [comment]
Hi, I'm not sure when I would be able to get a proper tuning of this done and think it'll be a good idea to get this merged first, so that if anyone would like to try, can do so easily. Further tuning and changes needed could be done in a separate PR
Will open a PR to the Wiki for adding info about Rover soon!

BTW, if anyone would like to test it out without going through the hassle of installing AirSim and all, I've built binaries of AirSim with the Rover changes in Blocks environment, for both Linux and Windows, can be found here - https://drive.google.com/open?id=1lUoswRvjsNnC6_Wc1qgbc66DCp2UP09w

File: libraries/SITL/SIM_AirSim.cpp [hunk 2]
```diff
@@ -61,9 +70,9 @@ void AirSim::set_interface_ports(const char* address, const int port_in, const i
 /*
 	Decode and send servos
 */
-void AirSim::send_servos(const struct sitl_input &input)
+void AirSim::output_copter(const struct sitl_input &input)
 {
-	servo_packet pkt{0};
+    servo_packet pkt;
```
Comment (peterbarker) [inline libraries/SITL/SIM_AirSim.cpp]: I expect you've satisfied yourself this is pointless (and perhaps bad) as we definitely fill the packet?

File: libraries/SITL/SIM_AirSim.h [hunk 3]
```diff
@@ -43,12 +43,21 @@ class AirSim : public Aircraft {
     void set_interface_ports(const char* address, const int port_in, const int port_out) override;
 
 private:
-	/*
-		rotor control packet sent by Ardupilot
-	*/
-	static const int kArduCopterRotorControlCount = 11;
+    enum {
```
Comment (peterbarker) [inline libraries/SITL/SIM_AirSim.h]: `enum class OutputType`, please.

File: libraries/SITL/SIM_AirSim.cpp (hunk 2 above)
Comment (rajat2004) [reply libraries/SITL/SIM_AirSim.cpp]: Yeah, there were some pretty silly mistakes in the code :sweat_smile:

File: libraries/SITL/SIM_AirSim.h (hunk 3 above)
Comment (rajat2004) [reply libraries/SITL/SIM_AirSim.h]: Fixed, thanks for the review!

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**